File: Laboratory_2/Go/task2.go
package main

import "fmt"


func print_bottles(bottle string, count int) {
	/* Выводит бутылки на экран. */
	for i := 0; i < count; i++ {
		if i % 3 == 0 {
			fmt.Println()
		}
		fmt.Printf("%s ", bottle)
	}
}


func drink_simulation(n, m int) [2]int {
	/* Показывает все шаги питья. */
	full := n
	empty := 0
	total_drunk := 0
	iterations := 0

	full_bottle := "●"
	empty_bottle := "○"

	// Выведем дано
	fmt.Print("Дано")
	print_bottles(full_bottle, full)
	fmt.Print("\n\n")

	for {
		// Шаг 1: выпить все полные бутылки
		if full > 0 {
			total_drunk += full
			empty += full
			full = 0
			iterations++

			// Выведем все выпитые бутылки
			fmt.Printf("%d шаг", iterations)
			print_bottles(empty_bottle, empty)
			fmt.Print("\n\n")

			if (empty < m) {  // Условие завершения
				break
			}

			// Шаг 2: Обмен пустых бутылок на полные
			full = empty / m
			empty = empty % m
			iterations++

			// Выведем бутылки после обмена
			fmt.Printf("%d шаг", iterations)
			print_bottles(full_bottle, full)
			print_bottles(empty_bottle, empty)
			fmt.Print("\n\n")
		} 
	}

	return [2]int{total_drunk, iterations}
}


func task2() {
	var n, m int
	fmt.Print("Введите количество полных бутылок и количество бутылок, которых достаточно для обмена: ")
	fmt.Scan(&n, &m)

	fmt.Printf("Input: %d %d \n", n, m)

	result := drink_simulation(n, m)

	total := result[0]
	iters := result[1]

	fmt.Printf("Output (total, iters): %d %d \n", total, iters)
}

File: LaboratoryWork/Semester_2/Laboratory_2/Swift/task2.swift
import Foundation

func printBottles(bottle: String, count: Int) {
    /* Выводит бутылки на экран. */
    for i in 0..<count {
        if i % 3 == 0 {
            print()
        }
        print(bottle, terminator: " ")
    }
}

func drinkSimulation(n: Int, m: Int) -> (Int, Int) {
    /* Показывает все шаги питья. */
    var full = n
    var empty = 0
    var totalDrunk = 0
    var iterations = 0

    let fullBottle = "●"
    let emptyBottle = "○"

    // Выведем дано
    print("Дано", terminator: " ")
    printBottles(bottle: fullBottle, count: full)
    print("\n")

    while true {
        // Шаг 1: выпить все полные бутылки
        if full > 0 {
            totalDrunk += full
            empty += full
            full = 0
            iterations += 1

            // Выведем все выпитые бутылки
            print("\(iterations) шаг", terminator: " ")
            printBottles(bottle: emptyBottle, count: empty)
            print("\n")

            if empty < m { break } // условие завершения

            // Шаг 2: обмен пустых бутылок на полные
            full = empty / m
            empty = empty % m
            iterations += 1

            // Выведем бутылки после обмена
            print("\(iterations) шаг", terminator: " ")
            printBottles(bottle: fullBottle, count: full)
            printBottles(bottle: emptyBottle, count: empty)
            print("\n")
        }
    }

    return (totalDrunk, iterations)
}

// Основная часть программы
print("Введите количество полных бутылок и количество бутылок, которых достаточно для обмена: ", terminator: "")
if let input = readLine(), let n = Int(input.split(separator: " ")[0]), let m = Int(input.split(separator: " ")[1]) {
    print("Input: \(n) \(m)")

    let result = drinkSimulation(n: n, m: m)
    let total = result.0
    let iters = result.1

    print("Output (total, iters): \(total) \(iters)")
}
